File: 2_keys_keyboard.cpp
class Solution {
public:
    int minSteps(int n) {
        if (n == 1) return 0;
        
        int dp[n+1] = {};
        dp[0] = 0; dp[1] = 0;
        for (int i = 2; i <= n; ++i) {
            dp[i] = i;
            for (int j = 2; j < i; ++j) {
                if (i%j == 0) {
                    dp[i] = min(dp[i], dp[j] + 1 + (i/j -1));
                }
            }
        }
        return dp[n];
    }
};

File: 2_keys_keyboard.py
class Solution:
    def minSteps(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1: return 0
        
        dp = (n+1)*[None]
        for i in range(1, n+1):
            dp[i] = i
            for j in range(2,i):
                if i%j == 0:
                    dp[i] = min(dp[i], dp[j] + 1 + (i/j-1))
                    
        return int(dp[n])
        
